metals/src/main/scala/scala/meta/internal/metals/WorkspaceFolders.scala

```diff
@@ -35,32 +53,68 @@ class WorkspaceFolders(
   ): Future[Unit] = {
     val actualToRemove =
       toRemove.filterNot(folder => toAdd.exists(_.path == folder.path))
+
     def shouldBeRemoved(folder: Folder) =
       actualToRemove.exists(_.path == folder.path)
 
-    val (newScala, newNonScala) = toAdd.partition(_.isMetalsProject)
-    val newServices = newScala.map(createService(_))
-    if (newServices.isEmpty && getFolderServices.forall(shouldBeRemoved)) {
+    val WorkspaceFoldersServices(prev, prevDel, _) =
+      folderServices.getAndUpdate {
+        case WorkspaceFoldersServices(
+              services,
+              delegating,
+              nonScalaProjects,
+            ) =>
+          val filteredServices = services.filterNot(shouldBeRemoved)
+          val filteredDelegating = delegating.filterNot(shouldBeRemoved)
+
+          val (newScala, newNonScala) = toAdd
+            .filterNot(isIn(services ++ delegating ++ nonScalaProjects, _))
+            .partition(_.isMetalsProject)
+
+          val allNewScala = newScala.map { folder =>
+            findDelegate(filteredServices, folder) match {
+              case Some(service) =>
+                DelegatingFolderService(folder, service)
+              case None => createService(folder)
+            }
+          }
+
+          val updatedServices = {
+            val transformedDelegating =
+              filteredDelegating.collect {
+                case del if shouldBeRemoved(del.service) => createService(del)
```

Comment:
We used to `createService` outside of `getAndUpdate` but I don't think that just creating an instance has any side effects, so this should be safe.